### PR TITLE
[DNM]cmake: detect C compiler SIMD build option

### PR DESCRIPTION
**Pull request by changchengx**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by wjwithagen** [comment]
I wonder how this works for building packages and/or rpms...
Suppose a package is build on the newest CPU with all bells and whistles.
That 'll select AVX stuff.
But th table below suggest that there will be architecture series that will not be able to run this code...
Unless GCC or Clang include dual code: with avx and without avx?

Flag | Description
-- | --
-march=native | Creates an executable that uses SIMD instructions based on the CPU  that is compiling the code. Additionally it includes the optimizations  from the -mtune=native flag.
-march=arch | This will generate SIMD instructions for a particular architecture and apply the -mtune optimizations.  The useful values of arch are the same as for the -mtune flag above.
-msse4.2 | Generates code with SSE4.2 instructions.
-mavx | Generates code with AVX instructions. Code compiled with this flag will not be able to run on Nehalem architecture cores.
-mavx2 | Generates code with AVX2 instructions. This requires an additional  module to be loaded before compiling: module load binutils/2.28 Code compiled with this flag will not be able to run on Nehalem, Sandybridge, Ivybridge, or the AMD Bulldozer cores. Before using this option, load a newer version of the linker program that will properly handle AVX2 instructions: module load binutils/2.28

**Comment by changchengx** [comment]
@wjwithagen 
1. This only detects whether the C compiler supports "-mavx2" or "-mavx512f" options.
Note: **They're not added into "SIMD_COMPILE_FLAGS" variables.**
2. There's another PR https://github.com/ceph/ceph/pull/35102 to detect whether the running hardware support AVX2 or AVX512 related instructions

So, mainly:
1. Check whether C compiler support "-mavx2" or "-mavx512f"
2. runtime check whether the running hardware support arch-specific instructions

**Comment by changchengx** [comment]
jenkins, test this please

File: cmake/modules/SIMDExt.cmake [hunk 1]
```diff
@@ -77,6 +77,10 @@ elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i386|i686|amd64|x86_64|AMD64")
       if(HAVE_INTEL_SSE4_2)
         set(SIMD_COMPILE_FLAGS "${SIMD_COMPILE_FLAGS} -msse4.2")
       endif()
+      CHECK_C_COMPILER_FLAG(-maxv2 HAVE_INTEL_AVX2)
+      if(HAVE_INTEL_AVX2)
+        set(AVX2_COMPILE_FLAGS "-mavx2")
```
Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake:82]: how would `AVX2_COMPILE_FLAGS` be used?

File: cmake/modules/SIMDExt.cmake [hunk 2]
```diff
@@ -81,6 +81,10 @@ elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i386|i686|amd64|x86_64|AMD64")
       if(HAVE_INTEL_AVX2)
         set(AVX2_COMPILE_FLAGS "-mavx2")
       endif()
+      CHECK_C_COMPILER_FLAG(-mavx512f HAVE_INTEL_AVX512F)
+      if(HAVE_INTEL_AVX512F)
+        set(AVX512F_COMPILE_FLAGS "-mavx512f")
```
Comment (tchaikov) [inline cmake/modules/SIMDExt.cmake:86]: ditto.

File: cmake/modules/SIMDExt.cmake (hunk 1 above)
Comment (changchengx) [reply cmake/modules/SIMDExt.cmake:82]: "AVX2_COMPILE_FLAGS" should have similar usage as "SIMD_COMPILE_FLAGS". When building some libraries in future, we could build it with "-mavx2"(AVX2_COMPILE_FLAGS) options. At runtime, there'll be some other mechanisms to check whether the hardware support AVX2 or not.
Currently, "SIMD_COMPILE_FLAGS" includes too many options, such as "-msse2" & "-mssse3" and so on. If we add "-mavx2" into SIMD_COMPILE_FLAGS, there'll be some unknown stability problems in jerasure.

File: cmake/modules/SIMDExt.cmake (hunk 2 above)
Comment (changchengx) [reply cmake/modules/SIMDExt.cmake:86]: @tchaikov Same as above.

File: cmake/modules/SIMDExt.cmake (hunk 1 above)
Comment (tchaikov) [reply cmake/modules/SIMDExt.cmake:82]: as discussed, i suggest use MFV (https://lwn.net/Articles/691932/) to implement this.

Comment (changchengx) [reply cmake/modules/SIMDExt.cmake:82]: I'll check it later

**Comment by wjwithagen** [comment]
> 
> 
> @wjwithagen
> 
>     1. This only detects whether the C compiler supports "-mavx2" or "-mavx512f" options.
>        Note: **They're not added into "SIMD_COMPILE_FLAGS" variables.**
> 
>     2. There's another PR #35102 to detect whether the running hardware support AVX2 or AVX512 related instructions
> 
> 
> So, mainly:
> 
>     1. Check whether C compiler support "-mavx2" or "-mavx512f"
> 
>     2. runtime check whether the running hardware support arch-specific instructions

Right, in combintation with the FMV options,  that make this an interesting enhancement.
@tchaikov thanx for the pointer to that FMV article, Clang has the same options, and I tested the example which seems to work.

**Comment by changchengx** [comment]
@tchaikov 
FMV doesn't work well.
For example ```__attribute__ ((target ("avx512cd")))``` or ```__attribute__ ((target("avx512pf"))``` and so on

**Comment by tchaikov** [comment]
@changchengx please define "doesn't work well". see https://gcc.gnu.org/onlinedocs/gcc-9.1.0/gcc/x86-Function-Attributes.html and https://patchwork.ozlabs.org/project/gcc/patch/CAMXFM3uzV8nTWf-UHrmpyU8Qkt72Nv1+UiUzKJCb1hc+VXww+w@mail.gmail.com/ and https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/avx512cdintrin.h

 i believe GCC-9.1 should support, for instance `target("avx512cd")`.

**Comment by changchengx** [comment]
@tchaikov 
```
nstcc4@nstcloudcc4:~$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

nstcc4@nstcloudcc4:~$ cat -n test.cc
     1  __attribute__ ((target("avx512cd")))
     2  int foo(){
     3      return 2;
     4  }
     5  
     6  __attribute__ ((target("default")))
     7  int foo() {
     8      return 3;
     9  }
    10  
    11  int main() {
    12      foo();
    13      return 0;
    14  }
nstcc4@nstcloudcc4:~$ g++ test.cc -o go
test.cc: In function ‘_Z3foov.resolver’:
test.cc:2:5: error: ISA ‘avx512cd’ is not supported in ‘target’ attribute, use ‘arch=’ syntax
    2 | int foo(){
      |     ^~~
```

**Comment by tchaikov** [comment]
@changchengx please try:
 ```sh
g++ -mavx512cd -O0 -g test.cc -o test
```

**Comment by changchengx** [comment]
@tchaikov 
Let's hold this PR until I verify it locally that it could work in Ceph.
It may take a little long time because I need to focus on one feature development until the end of December.

**Comment by changchengx** [comment]
Refresh, hold it here

**Comment by LiucIntel** [comment]
@majianpeng @hualongfeng 
You can continue check this PR.
Note: MFV feature is useful. However, I can't confirm whether it will also including checking whether kernel disable AVX context.

**Comment by majianpeng** [comment]
Got it.

Thanks!
Jianpeng.

________________________________________

**Comment by changchengx** [comment]
Close this PR.
Intel employees could submit new PR if it's needed.

**Pull request closed without merging**